### ISSUE TITLE
Do not set Content-Type header for requests with no body

### ISSUE THREAD

**Issue by kminder** [rest-assured/rest-assured#656]
I recently upgraded from a rather old version of rest-assured.  I had to make several adjustments but one that I didn't think was particularly valid was needing to "clear" a content type for a PUT without a body.  Note as shown below that if I don't do a contentType(""), then rest-assured will now send a default Content-Type: application/json.  Having a default is fine but when there is no body there should probably be no content type sent.

```
given()
    .auth().preemptive().basic( username, password )
    .header( "X-XSRF-Header", "random-value" )
    .expect()
    .statusCode( HttpStatus.SC_CREATED )
    .contentType( ContentType.XML )
    .header( "Location", startsWith( driver.getUrl( "TESTROOT" ) + path ) )
    .when().contentType("").put( driver.getUrl( "TESTROOT" ) + path);
```

**Comment by johanhaleby** [comment]
True, if there's no content we shouldn't add a content-type header. Thanks for reporting. Care to provide a pull request for this?

**Comment by pointwisefixed** [comment]
Hi!, this doesn't happen to me in my test cases. Could you provide a simple test case that makes the content type set to json?

**Comment by johanhaleby** [comment]
Thanks for trying it our @pointwisefixed

**Comment by johanhaleby** [comment]
I've written a test case for this:

```java
@Test public void
doesnt_send_a_content_type_header_when_there_is_no_body() {
    when().
            get("/headersWithValues").
    then().
            body("containsKey('Content-Type')", is(false));
}
```

where "/headersWithValues" is implemented like this:

```scala
get("/headersWithValues") {
    contentType = "application/json"
    val headerNames = request.getHeaderNames.map(_.toString)

    val map: Map[String, List[String]] = headerNames.map(headerName => (headerName, request.getHeaders(headerName).map(_.toString).toList)).
      foldLeft(mutable.HashMap[String, List[String]]())((map, header) => {
      map.put(header._1, header._2.toList)
      map
    }).toMap
    compact(render(decompose(map)))
}	
```

And the test pass. So it doesn't seem like this is an issue (at least I can't reproduce it).

**Comment by muralinadadhur** [comment]
Hi, My test goes like this 

`ExtractableResponse<Response> spec = given()
                .headers(hBuilder.build())
                .expect()
                .statusCode(isOneOf(201, 409))
                .request()
                .when().log().parameters().log().headers()
                .post(path + EndpointsRestSdk.FILE_RESUMABLE_SUFFIX)
                .then().log().all()
                .extract();`

I'm not passing any Content-Type since there's no content here, but the request output is setup like this - `Request params:	<none>
Query params:	<none>
Form params:	<none>
Path params:	<none>
Multiparts:		<none>
Headers:		Authorization=Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik1VWkdNVEJFTVRsQk56Z3hNVEF4UmtKRVFUVkdOMEZHTkVVNE1EUTJNVE15TUVORE9VSkNOUSJ9.eyJzY29wZXMiOiJyZWFkLW9ubHkgcmVhZC13cml0ZSBuYXNfcmVhZF9vbmx5IG5hc19yZWFkX3dyaXRlIG5hc19vZmZsaW5lX2FjY2VzcyBuYXNfYXBwX21hbmFnZW1lbnQgZGV2aWNlX2dldCBkZXZpY2VfZ2V0X3VzZXJfZGV2aWNlcyBkZXZpY2VfYXR0YWNoIGRldmljZV9kZXRhY2ggZGV2aWNlX2dldF9hcHByb3ZhbHMgZGV2aWNlX2dldF9kZXZpY2VfdXNlcnMgZGV2aWNlX2FwcHJvdmVfdXNlcnMgZGV2aWNlX2ludml0ZV91c2VycyBkZXZpY2VfZ2V0X2ludml0YXRpb24gZGV2aWNlX2dldF91c2VyIG90YV9yZWFkX29ubHkiLCJpc3MiOiJodHRwczovL3dkYy1xYTEuYXV0aDAuY29tLyIsInN1YiI6ImF1dGgwfDU5YjBhMTQyYmYyMDhmNzA3ZTFlNTUwNCIsImF1ZCI6Im1RdEFnV2x1b1ladE1VNTdwbWs1NWdOYklBc3BxdW91IiwiZXhwIjoxNTA2NTQ5ODM5LCJpYXQiOjE1MDY0NjM0Mzl9.81zterWwkKQnS8ncfm-tGRIBe7yOG8FOj3iY2xZFcOrsPXiZSctdcMa0IrBMXSnOF-__nT9FIQEEId1mASPlx8UJjl9xDn1cvSWzig18cbtKVXXKVwNRgMdQkINV0x_Fpo6u4IK4Sy1Nxwgkcy4iVgVps1fuLA4BJxshzx7d51BmQ7gOsBSFx4ULW8QtVSxtdNjLuVsbEJ3xP2KmEpUoMtkOsWgt7BDc_Kii39rTTHpIga3RjZYXvvNePYRs4v15brS-5usd0Ezs6qt-ZibNBoAVI8L6dj533O2AMWQ5xtKvCWoyDE19KUvOR6ro8a893jRXOIhxV-nLo_9euIOFSg
				Accept=*/*
				Content-Type=application/x-www-form-urlencoded; charset=ISO-8859-1HTTP/1.1 201 Created`

**Comment by johanhaleby** [comment]
What version are you using?

**Comment by srikar02** [comment]
@johanhaleby 

I am using version 3.0.6 and I can replicate this issue. As you can notice I am not passing in any body. Yet rest assured adds the content type. My test fails as I am expecting a 401 but instead getting 415

Code
```java
//Validate 401

		given().baseUri(props.getProperty("BaseUri")).pathParam("storeid", props.getProperty("storeid")).log().all().

				when().request("POST", "/stores/{storeid}/tickets").prettyPeek().

				then().assertThat().statusCode(401).and().contentType(ContentType.HTML)
				.statusLine("HTTP/1.1 401 Unauthorized");
```

Output
`
Request method:	POST
Request URI:	https://<Masked>/stores/12345/tickets
Proxy:			<none>
Request params:	<none>
Query params:	<none>
Form params:	<none>
Path params:	storeid=12345
Headers:		Accept=*/*
				Content-Type=application/x-www-form-urlencoded; charset=ISO-8859-1
Cookies:		<none>
Multiparts:		<none>
Body:			<none>
HTTP/1.1 415 Unsupported Media Type
`

My pom
`<dependency>
			<groupId>io.rest-assured</groupId>
			<artifactId>rest-assured</artifactId>
			<version>3.0.6</version>
			<scope>test</scope>
</dependency>`

**Comment by johanhaleby** [comment]
@srikar02 Thanks, that's a good example! Sounds reasonable that the content-type shouldn't be supplied for empty bodies (at least not be default).

**Comment by LorenzoBettini** [comment]
I'm experiencing the same problem: I cannot seem to test a POST end-point with a null object (corresponding to an empty body). You said that you managed to clear the content type with contentType("") but it's not working in my case: the header is still there (content-type=[; charset=ISO-8859-1]) and my Java method is not being called with a null object (my method is not called at all).

**Comment by avistech** [comment]
hi guys, may i know if this issue is fixed (or) someone found a workaround. This is causing confusion  when we publish API for third parties to consume. thanks

**Comment by kool79** [comment]
When this will nbe fixed?? it affects my application also. v4.3.3

**Comment by simpleDev24** [comment]
Faced same issue and was able to workaround by setting content type as empty string and also providing encoder config so that RA don't add default character set --

`RequestSpecification request = given()
                .config(RestAssured.config()
                        .encoderConfig(RestAssured.config().getEncoderConfig().appendDefaultContentCharsetToContentTypeIfUndefined(
                                false)))
                .filters(new RequestLoggingFilter(LogDetail.ALL, false, System.out),
                        new ResponseLoggingFilter(LogDetail.ALL, false, System.out))
                .accept("*/*")
                .contentType("");`

If you send request using above request specification then Content-Type header in request have blank value -- which worked in my case. This is how the header is sent in request-- 
`Content-Type=`
